File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    public TreeNode lca(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root.val == p.val || root.val == q.val) {
            return root;
        }

        if (p.val < root.val && q.val < root.val) {
            return lca(root.left, p, q);
        } else if (p.val > root.val && q.val > root.val) {
            return lca(root.right, p, q);
        }
        return root;
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return lca(root, p, q);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def recur(root):
            if not root:
                return [True, 0]
            isBL, hL = recur(root.left)
            isBR, hR = recur(root.right)
            isBalance = abs(hL - hR) <= 1
            maxH = 1 + max(hL, hR)
            return [isBL and isBR and isBalance, maxH]
        
        balance, _ = recur(root)
        return balance
